llvm/lib/CodeGen/SelectOptimize.cpp

```diff
@@ -227,72 +176,55 @@ class SelectOptimizeImpl {
         return Sel->getFalseValue();
       // Or(zext) case - return the operand which is not the zext.
       if (auto *BO = dyn_cast<BinaryOperator>(I)) {
-        Value *X;
-        if (PatternMatch::match(BO->getOperand(0),
-                                m_OneUse(m_ZExt(m_Value(X)))))
-          return BO->getOperand(1);
-        if (PatternMatch::match(BO->getOperand(1),
-                                m_OneUse(m_ZExt(m_Value(X)))))
-          return BO->getOperand(0);
+        // Return non-dependant on condition operand
```

Comment:
BinaryOperator is select-like instruction where one operand depends on a condition, with the index of that operand being `CondIdx`. `non-dependant on condition operand` is the other operand. On the branch where the condition is 0, BinaryOperator doesn't do anything, so we can just return the other operand.

I feel like matching condition-depending operand with ZExt was not entirely correct - for example, what if both operands are ZExts? And late, after adding `LShr` and `AShr` support, it would get even messier.